### ISSUE TITLE
Training from scratch for 2 classes, getting nan values

### ISSUE THREAD

**Issue by AnaRhisT94** [zzh8829/yolov3-tf2#75]
Hi @zzh8829 !
**Thanks for creating this awesome project.** 
I'm having difficulties for 2 days already to train from scratch my data. 
**Problem:** I'm receiving `nan` values when training in all the `loss` parameters. 

My `.txt` files look like this:
```
0 0.75078125 0.26481481481481484 0.05364583333333333 0.024074074074074074
1 0.7997395833333333 0.21898148148148147 0.2921875 0.43796296296296294
```
Where `0` is `plates` class, and `1` is vehicles class.

Then, In order to create my TFRecord files, I've written this code, debugged many times and it looks good:

My `create_tfrecord.py` file:

```
import os
import hashlib
import random
from pathlib import Path

import cv2 as cv
import tensorflow as tf
from absl import app


# train_txt = './train.txt'
# images_path = './JPEGImages'


def bytes_feature(value):
    return tf.train.Feature(bytes_list=tf.train.BytesList(value=[value]))


def bytes_list_feature(value):
    return tf.train.Feature(bytes_list=tf.train.BytesList(value=value))


def float_list_feature(value):
    return tf.train.Feature(float_list=tf.train.FloatList(value=value))


def int64_feature(value):
    return tf.train.Feature(int64_list=tf.train.Int64List(value=[value]))


def int64_list_feature(value):
    return tf.train.Feature(int64_list=tf.train.Int64List(value=value))


def get_example(line, image_path):
    class_text = []
    labels = []
    xmin = []
    ymin = []
    xmax = []
    ymax = []
    truncated = []
    poses = []
    difficult_obj = []

    # read the image
    with tf.io.gfile.GFile(image_path, 'rb') as fib:
        image_encoded = fib.read()
    h, w, _ = cv.imread(image_path).shape
    key = hashlib.sha256(image_encoded).hexdigest()

    # read the parameters
    item = line.split(' ')
    if item[0] == "0":
        # class 0 is plates
        class_text.append("plates".encode('utf8'))
        labels.append(0)
    elif item[0] == "1":
        # class 1 is vehicles
        class_text.append("vehicles".encode('utf8'))
        labels.append(1)
    else:
        raise ValueError("Error! I expect to get `0` for plates and `1` "
                         "for vehicles. I got {} instead.".format(item[0]))

    xmin.append(float(item[1]))
    ymin.append(float(item[2]))
    xmax.append(float(item[3]))
    ymax.append(float(item[4]))

    truncated.append(0)
    difficult_obj.append(0)
    poses.append("Unspecified".encode('utf8'))

    example = tf.train.Example(features=tf.train.Features(feature={
        'image/height': int64_feature(h),
        'image/width': int64_feature(w),
        'image/filename': bytes_feature(image_path.encode('utf8')),
        'image/source_id': bytes_feature(image_path.encode('utf8')),
        'image/key/sha256': bytes_feature(key.encode('utf8')),
        'image/encoded': bytes_feature(image_encoded),
        'image/format': bytes_feature('png'.encode('utf8')),
        'image/object/bbox/xmin': float_list_feature(xmin),
        'image/object/bbox/xmax': float_list_feature(xmax),
        'image/object/bbox/ymin': float_list_feature(ymin),
        'image/object/bbox/ymax': float_list_feature(ymax),
        'image/object/class/text': bytes_list_feature(class_text),
        'image/object/class/label': int64_list_feature(labels),
    }))

    return example


def create_tfrecord():
    # Create a TFRecordWriter, we will save the outputs in the following file
    train_writer = tf.io.TFRecordWriter('./tf_record_big/train.tfrecord')
    val_writer = tf.io.TFRecordWriter('./tf_record_big/val.tfrecord')

    # Counter to count the number of training and validation samples
    train_ctr = val_ctr = 0

    image_format_train = img_format_val = ".png"

    # list all the .txt files in the current folder (was previously ./mini_training_set)
    training_txt_files = [f for f in os.listdir('/home/me/Desktop/training_data/mini_training_set') if f.endswith('.txt')]
    txt_files_length = len(training_txt_files)
    # loop over all the training .txt files
    for i, txt_file in enumerate(training_txt_files):
        print("Working on file: {} / {}".format(i, txt_files_length))
        # get the associated image's file name and path
        image_name = txt_file.split(".")[0] + image_format_train
        image_path = os.path.join('/home/me/Desktop/training_data/mini_training_set', image_name)
        txt_path = os.path.join('/home/me/Desktop/training_data/mini_training_set', txt_file)
        if len(image_name) == 0 or len(image_path) == 0 or len(txt_path) == 0:
            continue
        if os.path.isfile(image_path) is False:
            print("Skipped that non-existant image.")
            continue
        with open(txt_path, 'r') as train_file:
            lines = train_file.read().split('\n')
            # random.shuffle(lines)
            # loop over all the training file lines
            for line in lines:
                if len(line) > 0:
                    example = get_example(line, image_path)
                    train_ctr += 1
                    train_writer.write(example.SerializeToString())
            train_file.close()
    print("Finishing reading the training .txt files.")
    print("Preparing to read now the validation .txt files.")

    # Prepare for creating TFRecord for validation set
    validation_txt_files = [f for f in os.listdir('/home/me/Desktop/training_data/mini_validation_set_same_as_training') if f.endswith('.txt')]
    val_files_length = len(validation_txt_files)
    # loop over all the training .txt files
    for i, txt_file in enumerate(validation_txt_files):
        print("Working on file: {} / {}".format(i, val_files_length))
        # get the associated image's file name and path
        image_name = txt_file.split(".")[0] + img_format_val
        if len(image_name) == 0 or len(image_path) == 0 or len(txt_path) == 0:
            continue
        image_path = os.path.join('/home/me/Desktop/training_data/mini_validation_set_same_as_training', image_name)
        txt_path = os.path.join('/home/me/Desktop/training_data/mini_validation_set_same_as_training', txt_file)
        if os.path.isfile(image_path) is False:
            print("Skipped that non-existant image.")
            continue
        with open(txt_path, 'r') as val_file:
            lines = val_file.read().split('\n')
            # random.shuffle(lines)
            # loop over all the training file lines
            for line in lines:
                if len(line) > 0:
                    example = get_example(line, image_path)
                    val_ctr += 1
                    val_writer.write(example.SerializeToString())
            val_file.close()

    print("Length of validation set: {}".format(val_ctr))
    print("Length of training set: {}".format(train_ctr))
    print("Finished reading all text files!")
    return True


if __name__ == '__main__':
    try:
        app.run(create_tfrecord())
    except SystemExit:
        pass
```

Finally, in the `database.py` I have changed the `labels` variable:
Before:
`labels = tf.cast(class_table.lookup(class_text), tf.float32)
`
After:
`labels = tf.cast(tf.sparse.to_dense(x['image/object/class/label'], default_value=0), tf.float32`

In `train.py` my flags are as follows:

```
flags.DEFINE_integer('size', 416, 'image size')
flags.DEFINE_integer('epochs', 10, 'number of epochs')
flags.DEFINE_integer('batch_size', 8, 'batch size')
flags.DEFINE_float('learning_rate', 1e-4, 'learning rate')
flags.DEFINE_integer('num_classes', 2, 'number of classes in the model')
```

**Comment by jllarraz** [comment]
What's your question?

**Comment by AnaRhisT94** [comment]
> What's your question?
@jllarraz 
During training I'm receiving nan values in all the loss parameters

**Comment by jllarraz** [comment]
If I figure out I will let you know

**Comment by seakkas** [comment]
I was getting nan values as well. I saw some \n values inside my tfrecord files. 

```bash
  feature {
    key: "image/object/class/text"
    value {
      bytes_list {
        value: "car\n"
        value: "car\n"
      }
    }
  }
```
After cleaning those \n (removed \ns in the annotation XML files), I don't see nan for now. 

Your tfrecord file might have similar issues.

To check the tfrecord file content:

```bash
import tensorflow as tf

for example in tf.python_io.tf_record_iterator("train.tfrecord"):
    file.write(str(tf.train.Example.FromString(example)))
```

p.s. I used ```create_pascal_tf_record.py``` to create tfrecord files.

**Comment by AnaRhisT94** [comment]
@seakkas  Thanks Selahattin! I think this is the last thing I need to verify, I'll update you tomorrow with my findings!

**Comment by AnaRhisT94** [comment]
@seakkas  Hi, the TFRecords still look good. were you able to train with TFRecords files?

**Comment by seakkas** [comment]
@AnaRhisT94  Yes, I was able to train it.

**Comment by AnaRhisT94** [comment]
@jllarraz  @seakkas 
Well, finally I'm able to get rid out of `nan` values.
The annotations I had did match the `darknet` ones (in terms of xmin,ymin, xmax, ymax) - the labeles,
but these labels didn't match when it had to be trained on this model.

I changed one ground truth labels until I was able to get rid out of `nans`, now I'm hoping it will give me the desired results. Afterwards I'll check for more than one image with labels and hopefully it will work on the other ones too as intended.

Will update soon.
